engine/protocol/api.go

```diff
@@ -0,0 +1,164 @@
+	headers storage.Headers,
+	executionResults storage.ExecutionResults,
+) API {
+	b := &backend{
+
+		headers:          headers,
+		blocks:           blocks,
+		state:            state,
+		executionResults: executionResults,
+	}
+	return b
+}
+
+func (b *backend) GetLatestBlock(_ context.Context, isSealed bool) (*flow.Block, error) {
+	var header *flow.Header
+	var err error
+
+	if isSealed {
+		// get the latest seal header from storage
+		header, err = b.state.Sealed().Head()
+	} else {
+		// get the finalized header from state
+		header, err = b.state.Final().Head()
+	}
+
+	if err != nil {
+		err = convertStorageError(err)
+		return nil, err
+	}
+
+	block, err := b.blocks.ByID(header.ID())
+	if err != nil {
+		err = convertStorageError(err)
+		return nil, err
+	}
+
+	return block, nil
+}
+
+func (b *backend) GetBlockByID(_ context.Context, id flow.Identifier) (*flow.Block, error) {
+	block, err := b.blocks.ByID(id)
+	if err != nil {
+		err = convertStorageError(err)
+		return nil, err
+	}
+
+	return block, nil
+}
+
+func (b *backend) GetBlockByHeight(_ context.Context, height uint64) (*flow.Block, error) {
+	block, err := b.blocks.ByHeight(height)
+	if err != nil {
+		err = convertStorageError(err)
+		return nil, err
+	}
+
+	return block, nil
+}
+
+func (b *backend) GetLatestBlockHeader(_ context.Context, isSealed bool) (*flow.Header, error) {
+	var header *flow.Header
+	var err error
+
+	if isSealed {
+		// get the latest seal header from storage
+		header, err = b.state.Sealed().Head()
+	} else {
+		// get the finalized header from state
+		header, err = b.state.Final().Head()
+	}
+
+	if err != nil {
+		err = convertStorageError(err)
+		return nil, err
+	}
+
+	return header, nil
+}
+
+func (b *backend) GetBlockHeaderByID(_ context.Context, id flow.Identifier) (*flow.Header, error) {
+	header, err := b.headers.ByBlockID(id)
+	if err != nil {
+		err = convertStorageError(err)
+		return nil, err
+	}
+
+	return header, nil
+}
+
+func (b *backend) GetBlockHeaderByHeight(_ context.Context, height uint64) (*flow.Header, error) {
+	header, err := b.headers.ByHeight(height)
+	if err != nil {
+		err = convertStorageError(err)
+		return nil, err
+	}
+
+	return header, nil
+}
+
+// GetExecutionResultByID gets an execution result by its ID.
+func (b *backend) GetExecutionResultByID(ctx context.Context, id flow.Identifier) (*flow.ExecutionResult, error) {
+	result, err := b.executionResults.ByID(id)
+	if err != nil {
+		return nil, convertStorageError(err)
+	}
+
+	return result, nil
+}
+
+func (b *backend) GetExecutionResultByBlockID(ctx context.Context, blockID flow.Identifier) (*flow.ExecutionResult, error) {
+	executionResult, err := b.executionResults.ByBlockID(blockID)
+	if err != nil {
+		return nil, convertStorageError(err)
+	}
+
+	return executionResult, nil
+}
+
+func convertStorageError(err error) error {
+	if status.Code(err) == codes.NotFound {
+		// Already converted
+		return err
+	}
+	if errors.Is(err, storage.ErrNotFound) {
```

Comment:
Sometimes it is more straightforward and less code just to propagate the error. It is up to you.

Reply:
Thanks - I agree it would would probably be a cleaner, but since this code is used elsewhere in flow-go in this format, I'm going to keep it in case they had a specific reason for doing it that way.